aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -19,77 +19,62 @@ namespace native {
+  return std::tuple<Tensor, Tensor>(num_exchanges.mul_(-2).add_(1),
+                                    lu.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1));
 }
 
 Tensor det(const Tensor& self) {
-  TORCH_CHECK(at::isFloatingType(self.scalar_type()) &&
-           self.dim() == 2 && self.size(0) == self.size(1),
-           "det(", self.type(), "{", self.sizes(), "}): expected a 2D square tensor "
-           "of floating types");
-  double det_P;
-  Tensor diag_U;
-  int info;
-  std::tie(det_P, diag_U, info) = _lu_det_P_diag_U_info(self);
-  if (info > 0) {
-    return at::zeros({}, self.options());
-  } else {
-    return diag_U.prod().mul_(det_P);
-  }
+  squareCheckInputs(self);
+  TORCH_CHECK(at::isFloatingType(self.scalar_type()), "Expected a floating point tensor as input");
+
+  Tensor det_P, diag_U;
+  std::tie(det_P, diag_U) = _lu_det_P_diag_U(self);
+  // complete_det is 0 when U is singular (U(i, i) = 0 for some i in [1, self.size(-1)]).
+  // The product accumulation takes care of this case, and hence no special case handling is required.
+  auto complete_det = diag_U.prod(-1).mul_(det_P);
+  return complete_det;
 }
 
 Tensor logdet(const Tensor& self) {
-  TORCH_CHECK(at::isFloatingType(self.scalar_type()) &&
-           self.dim() == 2 && self.size(0) == self.size(1),
-           "logdet(", self.type(), "{", self.sizes(), "}): expected a 2D square tensor "
-           "of floating types");
-  double det_P;
-  Tensor diag_U;
-  int info;
-  std::tie(det_P, diag_U, info) = _lu_det_P_diag_U_info(self);
-  if (info > 0) {
-    return at::full({}, -std::numeric_limits<double>::infinity(), self.options());
-  }
-  // `det_sign` is the sign of the determinant. We work on `diag_U.sign()` for
-  // numerical stability when diag_U has a lot small values.
-  auto det_sign = diag_U.sign().prod().mul_(det_P);
-  // This synchronizes on GPU, but `_lu_det_P_diag_U_info` above already synchronizes
-  if (det_sign.item<double>() <= 0) {
-    return det_sign.log_();  // get proper nan (det<0) or -inf (det=0)
+  squareCheckInputs(self);
+  TORCH_CHECK(at::isFloatingType(self.scalar_type()), "Expected a floating point tensor as input");
+
+  Tensor det_P, diag_U;
+  std::tie(det_P, diag_U) = _lu_det_P_diag_U(self);
+  Tensor det_sign = diag_U.sign().prod(-1).mul_(det_P);
+
+  // If det_sign > 0, diag_U.abs_().log_().sum(-1) gives logdet (this means U is not singular).
+  // If det_sign <= 0, then we get proper nan (when det < 0, i.e., det_sign) or -inf (when det = 0, i.e., U is singular).
+  // U is singular when U(i, i) = 0 for some i in [1, self.size(-1)].
+  Tensor logdet_vals = diag_U.abs_().log_().sum(-1);
+  if (self.dim() > 2) {
+    logdet_vals.index_put_((det_sign < 0).nonzero_numpy(), at::full({}, NAN, self.options()));
   } else {
-    return diag_U.abs_().log_().sum();
+    if ((det_sign < 0).item<uint8_t>()) {
```

Comment:
Can you instead `.item` first? On GPU this is another kernel launch for a tiny tensor.